### PR TITLE
Rework the bot

### PR DESCRIPTION
**Pull request by Bobby-McBobface**

### CODE REVIEW

**Comment by Awakened-Redstone** [comment]
Please note it uses .env

1st run .env example:
```
TOKEN=bot_token
TWITCH_CLIENT_ID=twitch_client_id
TWITCH_SECRET=twitch_secret
TWITCH_AUTH_TOKEN=
```
The twitch auth token will be filled automatically

**Comment by Awakened-Redstone** [comment]
Please test the modular levels before merging this into master

File: src/main/levels.py [hunk 1]
```diff
@@ -68,37 +69,50 @@ async def xp_needed_for_level(level: int):
...
+    for rank in ranks:
+        if level - 1 >= rank[1]:
+            try:
+                await member.add_roles(member.guild.get_role(rank[0]), reason="Level up!")
+                await member.remove_roles(member.guild.get_role(ranks[index + 1][0]), reason="Level up!")
+            except:
+                # User probably has the role already
+                pass
+        index += 1
+
```
Comment (Awakened-Redstone) [inline src/main/levels.py:84]: Has to be tested

**Review by Awakened-Redstone — CHANGES_REQUESTED** [review]
Test if modular levels works, it was not tested and may not work properly.

File: src/main/command/leveling/leaderboards.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        msg = await message.edit(embed=embed)
+        await self.leaderboard_pages(msg, page, client)
+
+    async def leaderboard_pages(self, message: discord.Message, page: int, client: discord.Client):
+        change_to = ''
+
+        def check(reaction, user):
+            print('aaaaaaaaaaa')
+            if user.bot:
+                return False
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:36]:
```suggestion
            if user.id not in await reaction.users().flatten():
                return False
```

File: src/main/command/leveling/leaderboards.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+            valid = str(reaction.emoji) == "◀️" or str(reaction.emoji) == "▶️"
+            if not valid:
+                return False
+
+            emoji = reaction.emoji
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:42]:
```suggestion
            emoji = reaction.emoji
            
            valid = emoji == "◀️" or emoji == "▶️"
            if not valid:
                return False

        
```

File: src/main/command/other/help.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+            for _cmd in command_list:
+                cmd: Command = _cmd()
+                name = cmd.command
+                arguments = ''
+                for argument in cmd.parameters:
+                    if argument.required:
+                        arguments += f" <{argument.identifier}>"
+                    elif not argument.required:
+                        arguments += f" [{argument.identifier}]"
```
Comment (Bobby-McBobface) [inline src/main/command/other/help.py:34]:
```suggestion
                arguments = ''.join(f" <{argument.identifier}>" if argument.required else f" [{argument.identifier}]"
```

File: src/main/command/other/help.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                elif command_list[len(command_list) - 1] == _cmd:
+                    category_commands += f"`{main.config['prefix']}{name}{arguments}` | {cmd.description}\n"
+                    embed.add_field(name=cmd.category.value.friendlyName, value=category_commands, inline=False)
+                else:
+                    category_commands += f"`{main.config['prefix']}{name}{arguments}` | {cmd.description}\n"
+
+            await message.channel.send(embed=embed)
+        else:
+            def filter_commands(command_):
+                return command_().command == parameters or parameters in command_().alias
```
Comment (Bobby-McBobface) [inline src/main/command/other/help.py:61]: Defining a new one each run

File: src/main/command/other/help.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    def __init__(self):
+        super().__init__("help", "Shows a command info or the list of the commands", ["commands", "?"], [
+            Parameter("command")], category=Category.GENERAL)
+
+    # noinspection PyMethodParameters
+    async def execute(self, message: discord.Message, parameters: str, client: discord.Client):
+        command_list = Command.__subclasses__()
+        command_list.sort(key=main.sort_commands)
+
+        if not parameters or parameters == '':
```
Comment (Bobby-McBobface) [inline src/main/command/other/help.py:17]:
```suggestion
        if not parameters:
```
Empty string is false in python

File: src/main/command/leveling/leaderboards.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            asyncio.get_running_loop().create_task(reaction.remove(user))
+            if emoji == "◀️" and page > 0:
+                self.change_to = 'previous'
+            elif emoji == "▶️":
+                self.change_to = 'next'
+            return valid
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:48]:
```suggestion
            if emoji == "◀️" and page > 0:
                page += -1
            elif emoji == "▶️":
                page += 1
            return True
```

File: src/main/command/leveling/leaderboards.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+        await client.wait_for('reaction_add', timeout=60.0, check=check)
+        if change_to == 'next':
+            await self.change_page(message, page + 1, client)
+        elif change_to == 'previous':
+            await self.change_page(message, page - 1, client)
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:54]:
```suggestion
        await self.change_page(message, page, client)
```

File: src/main/command/leveling/leaderboards.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        data_list = sqlite_client.execute('''SELECT ID, LEVEL, XP FROM LEVELS ORDER BY XP DESC LIMIT 3 OFFSET :offset''', {"offset": offset}).fetchall()
+
+        lb_list = ''
+        for data in data_list:
+            user = data[0]
+            level = data[1]
+            total_xp = data[2]
+            lb_list += f"<@{user}> | Level: {level} | Total XP: {total_xp}\n"
+
+        embed = discord.Embed(title="Leaderboard", description=lb_list)
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:71]: Keep it dry

File: src/main/command/leveling/leaderboards.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        data_list = sqlite_client.execute('''SELECT ID, LEVEL, XP FROM LEVELS ORDER BY XP DESC LIMIT 3 OFFSET :offset''', {"offset": page}).fetchall()
+
+        lb_list = ''
+        for data in data_list:
+            user = data[0]
+            level = data[1]
+            total_xp = data[2]
+            lb_list += f"<@{user}> | Level: {level} | Total XP: {total_xp}\n"
+
+        embed = discord.Embed(title="Leaderboard", description=lb_list)
```
Comment (Bobby-McBobface) [inline src/main/command/leveling/leaderboards.py:25]: Keep it dry

File: src/main/listener/messageReactionListener.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    if user.bot:
+        return
+    emoji: Union[discord.Emoji, discord.PartialEmoji, str] = reaction.emoji
+    reaction.message: discord.Message = reaction.message
+    if emoji == "◀️":
+        await reaction.message.edit(content="aaaaaaaaaa")
+        await reaction.remove(user)
+    elif emoji == "▶️":
+        await reaction.message.edit(content="AAAAAAAAAAAAAAAAAAAAAAAAA")
+        await reaction.remove(user)
```
Comment (Bobby-McBobface) [inline src/main/listener/messageReactionListener.py:17]: Do not need this file

File: src/main/main.py [hunk 12]
```diff
@@ -146,16 +158,72 @@ async def on_message(self, message):
...
+                for role_id in new_class.required_permissions.requiredRoles:
+                    roles += self.get_guild(config['guildId']).get_role(role_id)
+                if new_class.required_permissions.permissions.is_subset(user_permissions) and set(roles).issubset(set(message.author.roles)):
+                    await new_class.execute(message, parameters, self)
+                else:
+                    await message.channel.send("You don't have permission to do this.")
+
+    async def on_reaction_add(self, reaction: discord.Reaction, user: Union[discord.Member, discord.User]):
+        from listener import messageReactionListener
+        await messageReactionListener.on_reaction_add(reaction, user)
```
Comment (Bobby-McBobface) [inline src/main/main.py:216]:
```suggestion
```

File: config.json [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  "messages": {
+    "stringsHug": [
+      "{hugger} hugs {target}",
+      "{hugger} gives a big hug to {target}",
+      "{hugger} 🫂 {target}"
+    ],
+    "welcomeMsg": "{} has joined the game!",
+    "farewellMsg": "{} has left the game."
+  }
+}
```
[the remark this answers is not in the document]
Comment (Neron-png) [reply config.json:63]: It uses camel case, it's fine as long as it is consistent

File: src/main/command/leveling/leaderboards.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+        msg = await message.edit(embed=embed)
+        await self.leaderboard_pages(msg, page, client)
+
+    async def leaderboard_pages(self, message: discord.Message, page: int, client: discord.Client):
+        change_to = ''
+
+        def check(reaction, user):
+            print('aaaaaaaaaaa')
```
Comment (Neron-png) [inline src/main/command/leveling/leaderboards.py:34]:
```suggestion
```
Better remove line 34 from the master version, as it's in for debug purposes and will spam the terminal.

File: src/main/listener/messageReactionListener.py (hunk 11 above)
Comment (Neron-png) [reply src/main/listener/messageReactionListener.py:17]: Is this just for debugging?

**Review by Neron-png — APPROVED** [review]
I suggest we run it on the test server before we roll this out

File: src/main/listener/messageReactionListener.py (hunk 11 above)
Comment (Awakened-Redstone) [reply src/main/listener/messageReactionListener.py:17]: was testing reaction listener

**Pull request merged**